### PR TITLE
Exec fix

### PR DESCRIPTION
**Pull request by rennergade**
This is mostly in `nacl_syscall_common.c`

This PR fixes two issues with the Exec family. The more major of which was the improper conversion of NaCl 2D arrays for argv and envp into system args.

It also fixes `execv` by letting us execute `NaClSysExecve` with a NULL environment.

### CODE REVIEW

File: src/trusted/service_runtime/nacl_syscall_common.c [hunk 1]
```diff
@@ -4307,7 +4324,9 @@ int32_t NaClSysExecve(struct NaClAppThread *natp, void *pathname, void *argv, vo
 }
 
 int32_t NaClSysExecv(struct NaClAppThread *natp, void *pathname, void *argv) {
-  return NaClSysExecve(natp, pathname, argv, 0);
+  struct NaClApp *nap = natp->nap;
+
```
Comment (jesings) [inline src/trusted/service_runtime/nacl_syscall_common.c]: "nap" seems unused here.

File: src/trusted/service_runtime/nacl_syscall_common.c [hunk 2]
```diff
@@ -4307,7 +4324,9 @@ int32_t NaClSysExecve(struct NaClAppThread *natp, void *pathname, void *argv, vo
 }
 
 int32_t NaClSysExecv(struct NaClAppThread *natp, void *pathname, void *argv) {
```
Comment (jesings) [inline src/trusted/service_runtime/nacl_syscall_common.c]: You might also want to change the ```void *```s in this signature for consistency and clarity.

File: src/trusted/service_runtime/nacl_syscall_common.h [hunk 3]
```diff
@@ -315,7 +315,7 @@ int32_t NaClSysTestCrash(struct NaClAppThread *natp, int crash_type);
 int32_t NaClSysPipe(struct NaClAppThread *natp, uint32_t *pipedes);
 int32_t NaClSysPipe2(struct NaClAppThread *natp, uint32_t *pipedes, int flags);
 int32_t NaClSysFork(struct NaClAppThread *natp);
-int32_t NaClSysExecve(struct NaClAppThread *natp, void* pathname, void* argv, void* envp);
+int32_t NaClSysExecve(struct NaClAppThread *natp, char const *path, char *const *argv, char *const *envp);
 int32_t NaClSysExecv(struct NaClAppThread *natp, void *pathname, void *argv);
```
Comment (jesings) [inline src/trusted/service_runtime/nacl_syscall_common.h]: Same as above, changing the types on this signature might be worthwhile.

File: src/trusted/service_runtime/nacl_syscall_handlers_gen.py [hunk 4]
```diff
@@ -246,7 +246,7 @@
     ('NACL_sys_lind_syscall', 'NaClSysLindSyscall', ['uint32_t callNum', 'uint32_t inNum', 'void *inArgs', 'uint32_t outNum', 'void *outArgs']),
     ('NACL_sys_fork', 'NaClSysFork', []),
     ('NACL_sys_execv', 'NaClSysExecv', ['void *path', 'void *argv']),
```
Comment (jesings) [inline src/trusted/service_runtime/nacl_syscall_handlers_gen.py:248]: Here as well

**Review by jesings — COMMENTED** [review]
Overall looks like a significant improvement to what was previously there, there are a few changes I'd make for consistency before this should be merged.

**Comment by rennergade** [comment]
Alright, these are good suggestions. I actually have to keep the signatures in nacl_syscall_handlers_gen.py void because it doesn't have a way to deal with string arrays, but the rest I'll fix up now.

**Comment by rennergade** [comment]
I believe that fixes your issues! Let me know and I'll merge this in.

**Comment by jesings** [comment]
Yep this all seems very sane. Feel free to merge whenever!

**Pull request merged**